drift/lib/src/runtime/manager/references.dart

```diff
@@ -0,0 +1,492 @@
+///   3) Return the groups with the listings
+///
+/// Manually this would look like:
+///
+/// ```dart
+/// final groups = await groups.get();
+/// final groupIds = groups.map((group)=> group.id);
+/// final users = await users.filter((f)=> f.group.id.isIn(groupIds)).get()
+/// final groupsWithUsers = groups.map((group)=>(group,users.where((user)=> user.group == group.id)));
+/// ```
+///
+///
+/// Arguments:
+///   - [typedResults] is the raw result of the query, together with [currentTable] we can read it's results
+///   - [doPrefetch] is a bool which will only perform the prefetch if true.
+///   - [managerFromTypedResult] is the equivalent of:
+///     ```dart
+///     final groups = await groups.withReferences().get();
+///     for (final (group, refs) in groups){
+///         /// Manager with a filter to only get the users of this group.
+///         refs.users;
+///     }
+///     ```
+///     What we do, is collect all the filters from all of these `refs.users` and
+///     combine them with an OR operator to create a query which gets all the users
+///     [managerFromTypedResult] is the function which turns a single `group` into `refs.users`
+///   - [referencedTable] is a `MultiTypedResultKey` which is to write the results to the `TypedResult` object,
+///     This same class will be used in the `BaseReferences` class to read from the `TypedResult`.
+///   - [referencedItemsForCurrentItem] is the callback which does the mapping.
+///     It is the equivalent of`users.where((user)=> user.group == group.id)`.
+///
+/// This function is used by the generated code and should not be used directly.
+Future<List<TypedResult>> typedResultsWithPrefetched<
+        $CurrentDataclass,
+        $CurrentTable extends Table,
+        $ReferencedDataclass,
+        $ReferencedTable extends Table>(
+    {required bool doPrefetch,
+    required ProcessedTableManager<
+                dynamic,
+                dynamic,
+                $ReferencedDataclass,
+                dynamic,
+                dynamic,
+                dynamic,
+                dynamic,
+                dynamic,
+                $ReferencedDataclass,
+                dynamic>
+            Function(TypedResult)
+        managerFromTypedResult,
+    required List<TypedResult> typedResults,
+    required TableInfo<$CurrentTable, $CurrentDataclass> currentTable,
+    required MultiTypedResultKey referencedTable,
+    required Iterable<$ReferencedDataclass> Function(
+            $CurrentDataclass item, List<$ReferencedDataclass> referencedItems)
+        referencedItemsForCurrentItem}) async {
+  if (!doPrefetch || typedResults.isEmpty) {
+    return typedResults;
+  } else {
+    final managers = typedResults.map(managerFromTypedResult);
+    // Combine all the referenced managers into 1 large query which will return all the
+    // referenced items in one go.
+    final manager = managers.reduce((value, element) => value._filter(
```

Comment:
Does this work if the referenced items are from multiple different tables or would `typedResultsWithPrefetched` be called multiple times in that case?

Reply:
It's called once per prefetch